server/src/main/java/io/crate/execution/ddl/tables/AddColumnRequest.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Licensed to Crate.io GmbH ("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.execution.ddl.tables;
+
+import io.crate.metadata.Reference;
+import io.crate.metadata.RelationName;
+import io.crate.sql.tree.CheckColumnConstraint;
+import org.elasticsearch.action.support.master.AcknowledgedRequest;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Contains all supported for ADD COLUMN configurations such as constraints,
+ * doc values enabled flag, column type properties (bit length, text analyzer e.t.c).
+ */
+public class AddColumnRequest extends AcknowledgedRequest<AddColumnRequest> {
+
+    private final RelationName relationName;
+    private final boolean isPartitioned;
+    private final Reference columnRef; // Contains isNullable and hasDocValues which reflects NOT-NULL and STORAGE WITH.
+    private final Map<String, Object> columnProperties = new LinkedHashMap<>();
```

Comment:
Do you think it would make sense to make the properties concrete?

Downside is that we'll have streaming handling for changes, but those should be rare and it reduces the memory footprint + makes it explicit what properties are available

Reply:
> > I thought the negative positions are only used in the request classes, but that when the change is applied to the cluster state, they get converted to positive positions before it is saved/published.
> 
> This was the intention. I will see if I missed something.

just in case, how I spot this: enabled logging and saw that [this](https://github.com/crate/crate/blob/master/server/src/main/java/org/elasticsearch/index/mapper/MapperService.java#L189-L193) line is printed on a **non-master** node.
output was like:
[[mkhacawolgoxc.t/Z5xrIx2RTP-_O6Ok47_96w]] parsed mapping, and got different sources
```
original:
{"default":{"dynamic":"true","_meta":{"primary_keys":["id"]},"properties":{"id":{"type":"integer","position":1},"name":{"type":"keyword","position":-1}}}}
parsed:
{"default":{"dynamic":"true","_meta":{"primary_keys":["id"]},"properties":{"id":{"type":"integer","position":1},"name":{"type":"keyword","position":2}}}}
```
where original with negative position is `incomingMappingSource`

Reply:
Replied in https://github.com/crate/crate/pull/12806#discussion_r957236878